draft/XXXX-configuration_handling/README.md

```diff
@@ -0,0 +1,202 @@
+
+## Proposals
+
+This RFC will discuss a few concrete proposals. One is the overall architcture
+by which the Tinker Engine runtime receives, processes, and submits
+configuration data to a protocol. Another is the method by which a configuration
+file can express variation in its parameters. A third is a fallback mechanism
+allowing programmatic configuration handling when the declarative methods are
+not sufficient for some use case.
+
+### Overall Architecture
+
+Currently, invoking the `tinker` CLI requires specifying a configuration file,
+but currently Tinker does nothing with this file. The new behavior will be for
+Tinker to load the specified file and parse it as YAML (bearing in mind that
+valid JSON is also valid YAML). If the file cannot be opened, or if the contents
+do not parse properly, Tinker will exit with an error condition.
+
+The YAML content describes a "meta-configuration" (see below) that requires
+processing to turn it into a sequence of configurations, one per experimental
+run. This sequence of configurations will be used to construct one protocol per
+configuration, running each to complete one trial of the experimental protocol.
+
+This will be accomplished by sending each configuration dictionary to the
+protocol initializer (so that client logic can dictate how the configuration is
+used to set the protocol up), and then also as an argument to the protocol's
+`run()` method (allowing client logic to make use of the configuration data to
+direct the protocol runtime).
+
+The design of the particular experiment will be captured in the variation of
+configuration values, as expressed via *declarative meta-configuration* or
+*programmatic configuration*.
+
+### Declarative Meta-Configuration
+
+In many cases, the user of Tinker Engine will want to run a protocol multiple
+times, changing some set of parameters in order to observe the effect those
+changes have on the results of the run. In this way, Tinker Engine supports
+computational experiments through the use of the scientific method.
+
+The concept of "some set of parameters" is intentionally abstract: this may mean
+anything from selecting a set of numeric values for a protocol setting to take
+on, to automatically marching linearly through a range of floating point values,
+to swapping out entire algorithmic implementations to compare different
+approaches. These types of variations will be expressible through specific
+meta-configuration syntax within the configuration file schema. Some examples
+follow.
+
+#### Combinatorial variation
+
+The most straightforward idea is to provide lists of values for each varying
+paramater to assume, and then to produce a stream of configurations visiting
+every ordered tuple of such values in their collective product space. For
+instance, the following hypothetical snippet of a configuration:
+
+```yaml
+a: 0
+b:
+  values: [1, 2, 3]
+c:
+  values: [4, 5]
+```
```

Comment:
This looks good and makes sense.  We should have some printouts to say how many configurations are happening (since their protocol won't actually know since it's just being invoked each time.

Reply:
Good idea. I have called this an "execution plan" in the past. The idea would be to have a dry-run command line flag that would just show what configurations it would run after analyzing the meta-configuration. This way, the user can verify that they didn't, for instance, combinatorially explode by accident, etc.